src/main/python/run_trecqa.py

```diff
@@ -5,219 +5,294 @@
 
-  return scored_candidates
+    return scored_candidates
 
 
 def load_data(fname):
-  questions = []
-  answers = {}
-  labels = {}
-  prev = ''
-  answer_count = 0
+    questions = []
+    answers = {}
+    labels = {}
+    prev = ''
+    answer_count = 0
 
-  with open(fname, 'r') as f:
-    for line in f:
-      line = line.strip()
-      qid_match = re.match('<QApairs id=\'(.*)\'>', line)
+    with open(fname, 'r') as f:
+        for line in f:
+            line = line.strip()
+            qid_match = re.match('<QApairs id=\'(.*)\'>', line)
 
-      if qid_match:
-        answer_count = 0
-        qid = qid_match.group(1)
+            if qid_match:
+                answer_count = 0
+                qid = qid_match.group(1)
 
-        if qid not in answers:
-          answers[qid] = []
+                if qid not in answers:
+                    answers[qid] = []
 
-        if qid not in labels:
-          labels[qid] = {}
+                if qid not in labels:
+                    labels[qid] = {}
 
-      if prev and prev.startswith('<question>'):
-        questions.append(line)
+            if prev and prev.startswith('<question>'):
+                questions.append(line)
 
-      label = re.match('^<(positive|negative)>', prev)
+            label = re.match('^<(positive|negative)>', prev)
 
-      if label:
-        label = label.group(1)
-        label = 1 if label == 'positive' else 0
-        answer = line.split('\t')
-        answer_count += 1
+            if label:
+                label = label.group(1)
+                label = 1 if label == 'positive' else 0
+                answer = line.lower().split('\t')
+                answer_count += 1
 
-        answer_id = 'Q{}-A{}'.format(qid, answer_count)
-        answers[qid].append((answer_id, answer, label))
-        labels[qid][answer_id] = label
-      prev = line
+                answer_id = 'Q{}-A{}'.format(qid, answer_count)
+                answers[qid].append((answer_id, answer, label))
+                labels[qid][answer_id] = label
+            prev = line
 
-  return questions, answers, labels
+    return questions, answers, labels
 
 
-def create_ground_truth(fname, answers):
-  with open(fname, 'w') as f:
-    for qid in answers.keys():
-      for answer in answers[qid]:
-        f.write('{} 0 {} {}\n'.format(qid, answer[0], answer[2]))
+def create_qrel_jaccard(fname, answers):
+    with open(fname, 'w') as f:
+        for qid in answers.keys():
+            for answer in answers[qid]:
+                f.write('{} 0 {} {}\n'.format(qid, answer[0], answer[2]))
 
+# evaluate by TREC QA patterns
+def eval_by_pattern(qid, candidates, pattern, out):
+    seen = set([])
 
-def get_precision(actual, predicted, k):
-  total_count = 0.0
-  hits = 0.0
+    for i, cand in enumerate(candidates):
+        this_candidate = " ".join(cand[0])
 
-  for qid in predicted.keys():
-    run_labels = predicted.copy()
+        hash_value = hashlib.md5(this_candidate.encode()).hexdigest()
```

Comment:
Add a comment about why you're using the md5 hash, I can think of a reason, but I don't know, and it looks funky.